File: backend/src/main/java/com/checklistboteco/backend/security/TokenService.java
package com.checklistboteco.backend.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.smallrye.config.ConfigMapping;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

@ApplicationScoped
public class TokenService {
    @ConfigMapping(prefix="checklist.jwt") public interface JwtConfig { String secret(); }
    public static class Payload { public String userId; public boolean isAdmin; public long expiresAt; }
    @Inject ObjectMapper mapper;
    @Inject JwtConfig config;
    public String issue(String userId, boolean admin) {
        try {
            var payload=new Payload(); payload.userId=userId; payload.isAdmin=admin; payload.expiresAt=System.currentTimeMillis()+86_400_000L;
            String body=Base64.getUrlEncoder().withoutPadding().encodeToString(mapper.writeValueAsBytes(payload));
            return body+"."+sign(body);
        } catch(Exception e) { throw new IllegalStateException(e); }
    }
    public Payload verify(String token) {
        try {
            String[] parts=token.split("\\.");
            if(parts.length!=2 || !constantTime(sign(parts[0]),parts[1])) return null;
            Payload payload=mapper.readValue(Base64.getUrlDecoder().decode(parts[0]),Payload.class);
            return payload.expiresAt>System.currentTimeMillis()?payload:null;
        } catch(Exception e) { return null; }
    }
    private String sign(String value) throws Exception {
        Mac mac=Mac.getInstance("HmacSHA256"); mac.init(new SecretKeySpec(config.secret().getBytes(StandardCharsets.UTF_8),"HmacSHA256"));
        return Base64.getUrlEncoder().withoutPadding().encodeToString(mac.doFinal(value.getBytes(StandardCharsets.UTF_8)));
    }
    private boolean constantTime(String left,String right) { return MessageDigest.isEqual(left.getBytes(StandardCharsets.UTF_8),right.getBytes(StandardCharsets.UTF_8)); }
}

File: backend/src/main/kotlin/com/checklistboteco/backend/security/TokenService.kt
package com.checklistboteco.backend.security

import java.nio.charset.StandardCharsets
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

class TokenService(
    private val secret: String,
    private val clock: () -> Long = { System.currentTimeMillis() },
    private val ttlMillis: Long = 24L * 60L * 60L * 1000L
) {
    private val json = Json { encodeDefaults = true }

    fun issue(userId: String, isAdmin: Boolean): String {
        val payload = TokenPayload(
            userId = userId,
            isAdmin = isAdmin,
            expiresAt = clock() + ttlMillis
        )
        val payloadPart = json.encodeToString(TokenPayload.serializer(), payload).base64Url()
        val signature = sign(payloadPart)
        return "$payloadPart.$signature"
    }

    fun verify(token: String): TokenPayload? {
        val parts = token.split(".")
        if (parts.size != 2) return null
        if (sign(parts[0]) != parts[1]) return null
        val payload = runCatching {
            json.decodeFromString(TokenPayload.serializer(), String(Base64.getUrlDecoder().decode(parts[0]), StandardCharsets.UTF_8))
        }.getOrNull() ?: return null
        return payload.takeIf { it.expiresAt > clock() }
    }

    private fun sign(value: String): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret.toByteArray(StandardCharsets.UTF_8), "HmacSHA256"))
        return Base64.getUrlEncoder().withoutPadding().encodeToString(mac.doFinal(value.toByteArray(StandardCharsets.UTF_8)))
    }

    private fun String.base64Url(): String {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(toByteArray(StandardCharsets.UTF_8))
    }
}

@Serializable
data class TokenPayload(
    val userId: String,
    val isAdmin: Boolean,
    val expiresAt: Long
)
